### PR TITLE
Pulse 2722 - Idle Traptor when no rules are assigned

### PR DESCRIPTION
**Pull request by rdempsey**
Traptors are now able to idle for a configurable amount of time until rules are assigned to them in Redis. Once rules are found, the idle ceases and collection begins.

And yes, there is a test for it!

### CODE REVIEW

**Comment by madisonb** [comment]
Does this work both from the transition from 0 to >1 and from >1 to 0?

**Comment by rdempsey** [comment]
@madisonb yes it will. Due to the location of the _wait_for_rules() method, each time the Traptor is told to restart it will perform the check. If the rule count dropped to 0 it will idle for the configured time until rules are once again assigned to it.

**Comment by rdempsey** [comment]
@madisonb any more questions on this? If not I'm going to merge it in.

**Comment by madisonb** [comment]
@rdempsey who reviewed your PR?

File: tests/test_traptor_offline.py [hunk 1]
```diff
@@ -323,4 +324,17 @@ def test_ensure_traptor_only_enriches_tweets(self, traptor, non_tweet_stream_mes
         for message in non_tweet_stream_messages:
             print(message)
             enriched_data = traptor._enrich_tweet(message)
-            assert enriched_data == message
\ No newline at end of file
+            assert enriched_data == message
+
+    def test_ensure_traptor_stays_alive_until_rules_are_found(self, traptor):
```
Comment (jasonrhaas) [inline tests/test_traptor_offline.py]: Can you explain how this tests the method?  I'm still trying to wrap my head around MagicMock, Pytest, etc.

**Comment by jasonrhaas** [comment]
@madisonb I looked at it.  Looks good, just trying to understand the unit test piece of it.

**Comment by rdempsey** [comment]
@jasonrhaas the unit test works like this:

`traptor._setup()` - run  the `_setup()` function

`traptor.rule_check_interval = 2` - set the rule check interval to 2 seconds

`traptor.logger.debug = MagicMock()` - Mock the logger for debug level

`empty_response = {}` - create an empty response object to use for testing

`good_response = {"Rule": "Fun"}` - create a good response object to use for testing

`traptor._get_redis_rules = MagicMock(side_effect=[empty_response, empty_response, good_response])` - use MagicMock to make the `_get_redis_rules` function return 3 responses in the order provided, so bad, bad and then good.

`traptor._wait_for_rules()` - call the new `_wait_for_rules()` function which will check Redis for Rules assigned to the running Traptor.

'assert traptor.logger.debug.call_count == 2` - assert that two debug statements were produced during the run of the `_wait_for_rules()` function.

**Pull request merged**